support-figma/extended-layout-plugin/src/shader.ts

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Utils from "./utils";
+
+const SHADER_PLUGIN_DATA_KEY = "shader";
+const SHADER_FALLBACK_COLOR_PLUGIN_DATA_KEY = "shaderFallbackColor";
+
+export const shaderMap: ReadonlyMap<string, string> = new Map([
+  ["cloudy_sky", __uiFiles__.cloudy_sky],
+  ["discrete_ocean", __uiFiles__.discrete_ocean],
+  ["fibonacci_sphere", __uiFiles__.fibonacci_sphere],
+  ["gradient", __uiFiles__.gradient],
+  ["julia", __uiFiles__.julia],
+  ["star", __uiFiles__.star],
+]);
+
+export function onSelectionChanged() {
+  let selection = figma.currentPage.selection;
+
+  // We don't support multiple selections.
+  if (!selection || selection.length != 1 || !selection[0]) {
+    figma.ui.postMessage({ msg: "shader-selection-cleared" });
+    return;
+  }
+
+  let nodeWithFills: MinimalFillsMixin = selection[0] as MinimalFillsMixin;
+  if (nodeWithFills) {
+    figma.ui.postMessage({
+      msg: "shader-selection",
+      nodeId: selection[0].id,
+      size: { width: selection[0].width, height: selection[0].height },
+    });
+  }
+}
+
+export async function insertImage(imageBytes: Uint8Array) {
+  Utils.dcLog("Insert image", imageBytes);
+  await figma.loadAllPagesAsync();
+
+  let selection = figma.currentPage.selection;
+
+  // We don't support multiple selections.
+  if (!selection || selection.length != 1 || !selection[0]) {
+    figma.notify(
+      "No selections or multiple selections. Please select 1 node that can have an image fill..."
+    );
+    return;
+  }
+
+  let nodeWithFills: MinimalFillsMixin = selection[0] as MinimalFillsMixin;
+  if (nodeWithFills) {
+    // Set the image as the only fill for the node.
+    const imageHash = figma.createImage(imageBytes).hash;
+    nodeWithFills.fills = [
+      {
+        type: "IMAGE",
+        imageHash: imageHash,
+        scaleMode: "FILL",
+      },
+    ];
+  } else {
+    figma.notify(
+      "Current selection doesn't support image fill. Please select 1 node that can have an image fill..."
+    );
+  }
+}
+
+export async function setShader(shader: string, shaderFallbackColor: string) {
+  Utils.dcLog("Set shader: \n", shaderFallbackColor);
+  await figma.loadAllPagesAsync();
+  let selection = figma.currentPage.selection;
+
+  // We don't support multiple selections.
+  if (!selection || selection.length != 1 || !selection[0]) {
+    figma.notify(
+      "No selections or multiple selections. Please select 1 node that can have an image fill..."
+    );
+    return;
+  }
+  let nodeWithFills: MinimalFillsMixin = selection[0] as MinimalFillsMixin;
+
+  if (nodeWithFills) {
+    if (shader) {
+      selection[0].setSharedPluginData(
+        Utils.SHARED_PLUGIN_NAMESPACE,
+        SHADER_PLUGIN_DATA_KEY,
+        shader
+      );
+    } else {
```

Comment:
Can this ever get called? I don't see how shader can be undefined, which brings up the question of how to clear the shader from the node?

Reply:
Not at the moment. I am planning to add the clear shader menu in a following pr.